### PR TITLE
[dynamo] graph break on const dict KeyError

### PR DESCRIPTION
**Pull request by williamwen42**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #125945
* #125943
* __->__ #125882

Fixes https://github.com/pytorch/pytorch/issues/125866

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

### CODE REVIEW

File: torch/_dynamo/variables/builtin.py [hunk 1]
```diff
@@ -787,27 +787,29 @@ def call_self_handler(tx, args, kwargs):
...
-                        tx,
-                        fn(
+                    try:
+                        res = fn(
                             *[x.as_python_constant() for x in args],
-                        ),
-                    )
+                        )
+                    except Exception as exc:
+                        unimplemented(f"constant fold exception: {repr(exc)}")
```
Comment (jansel) [inline torch/_dynamo/variables/builtin.py:795]: What error is this catching?  I worry this will be hiding bugs.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/builtin.py (hunk 1 above)
Comment (williamwen42) [reply torch/_dynamo/variables/builtin.py:795]: This is catching an error on from a python call - `fn` would be a builtin function. Idea is that if we get a python error when trying to constant fold, we bubble up the error to skip the frame, then we evaluate the frame as normal, which would result in the same error. In the case of the repro, we would be catching a `KeyError`.

Comment (jansel) [reply torch/_dynamo/variables/builtin.py:795]: Let's make catch case more narrow, perhaps by checking the message and warning if it is somthing else.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/builtin.py (hunk 1 above)
Comment (williamwen42) [reply torch/_dynamo/variables/builtin.py:795]: I think that any possible exception raised by a function in `_constant_fold_functions` would have to be caught, e.g. KeyError, ZeroDivisionError, AttributeError, ValueError, etc. We could determine a list of allowed exceptions (by going through the builtin constant fold op list and manually checking for exceptions that could be raised), but if we miss any, then tracing any code like below would crash:
```python
def fn(x):
    try:
        y = ... # builtin constant fold op with constants that raises an exception we missed
        # e.g. we forgot ZeroDivisionError for y = 1 / 0
    except Exception:
        y = 1
    return x + y
```

What kind of bugs do you think would be hidden? My intent here is that if running the builtin op results in an exception, we should skip the frame, then default evaluation will also raise the exception and run any exception handlers.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**